src/bindings/python/flux/job.py

```diff
@@ -85,3 +96,385 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
+        elif count < 1:
+            raise ValueError("count must be > 0")
+
+        # validate the string keys
+        for key in ["id", "unit", "label"]:
+            if key in res and not isinstance(res[key], six.string_types):
+                raise TypeError("{} must be a string".format(key))
+
+        # validate the 'exclusive' key
+        if "exclusive" in res:
+            if res["exclusive"] not in [True, False]:
+                raise TypeError("exclusive must be a boolean")
+
+        # validate that slots have a 'label'
+        if res["type"] == "slot" and "label" not in res:
+            raise ValueError("slots must have labels")
+
+    @staticmethod
+    def _validate_task(task):
+        if not isinstance(task, collectionsAbc.Mapping):
+            raise TypeError("task must be a mapping")
+
+        _validate_keys(["command", "slot", "count"], task.keys(), allow_additional=True)
+
+        if not isinstance(task["count"], collectionsAbc.Mapping):
+            raise TypeError("count must be a mapping")
+
+        if not isinstance(task["slot"], six.string_types):
+            raise TypeError("slot must be a string")
+
+        if "attributes" in task and not isinstance(
+            task["attributes"], collectionsAbc.Mapping
+        ):
+            raise TypeError("count must be a mapping")
+
+        command = task["command"]
+        if not (
+            (  # sequence of strings - N.B. also true for just a plain string
+                isinstance(command, collectionsAbc.Sequence)
+                and all(isinstance(x, six.string_types) for x in command)
+            )
+        ) or isinstance(command, six.string_types):
+            raise TypeError("command must be a list of strings")
+
+    @staticmethod
+    def _validate_attributes(attributes):
+        _validate_keys(["system", "user"], attributes.keys(), keys_optional=True)
+
+    @staticmethod
+    def _create_resource(res_type, count, with_child=None):
+        if with_child is not None and not isinstance(
+            with_child, collectionsAbc.Sequence
+        ):
+            raise TypeError("child resource must None or a sequence")
+        elif with_child is not None and isinstance(with_child, six.string_types):
+            raise TypeError("child resource must not be a string")
+        if not count > 0:
+            raise ValueError("resource count must be > 0")
+
+        res = {"type": res_type, "count": count}
+
+        if with_child:
+            res["with"] = with_child
+        return res
+
+    @classmethod
+    def _create_slot(cls, label, count, with_child):
+        slot = cls._create_resource("slot", count, with_child)
+        slot["label"] = label
+        return slot
+
+    @property
+    def duration(self):
+        try:
+            return self.jobspec["attributes"]["system"]["duration"]
+        except KeyError:
+            return None
+
+    @duration.setter
+    def duration(self, duration):
```

Comment:
Might not have to be in this PR, but python standard duration as an argument here might also be good?